02-peer-protocol.md

```diff
@@ -560,6 +1038,302 @@ ones it has sent so it can use them should they be requested by
 incoming HTLCs.  The recipient only need remember one, for use in
 `r` route hints in BOLT 11 invoices.
 
+If an RBF negotiation is in progress when a `channel_ready` message is
+exchanged, the negotiation must be abandoned.
+
+## Channel Establishment v2
+
+This is a revision of the channel establishment protocol.
+It changes the previous protocol to allow the `accept_channel2` peer
+(the *accepter*/*non-initiator*) to contribute inputs to the funding
+transaction, via the interactive transaction construction protocol.
+
+        +-------+                              +-------+
+        |       |--(1)--- open_channel2  ----->|       |
+        |       |<-(2)--- accept_channel2 -----|       |
+        |       |                              |       |
+    --->|       |      <tx collaboration>      |       |
+    |   |       |                              |       |
+    |   |       |--(3)--  commitment_signed -->|       |
+    |   |       |<-(4)--  commitment_signed ---|       |
+    |   |   A   |                              |   B   |
+    |   |       |<-(5)--  tx_signatures -------|       |
+    |   |       |--(6)--  tx_signatures ------>|       |
+    |   |       |                              |       |
+    |   |       |--(a)--- tx_init_rbf -------->|       |
+    ----|       |<-(b)--- tx_ack_rbf ----------|       |
+        |       |                              |       |
+        |       |--(7)--- channel_ready  ----->|       |
+        |       |<-(8)--- channel_ready  ------|       |
+        +-------+                              +-------+
+
+        - where node A is *opener*/*initiator* and node B is
+          *accepter*/*non-initiator*
+
+### The `open_channel2` Message
+
+This message initiates the v2 channel establishment workflow.
+
+1. type: 64 (`open_channel2`)
+2. data:
+   * [`chain_hash`:`chain_hash`]
+   * [`channel_id`:`zerod_channel_id`]
+   * [`u32`:`funding_feerate_perkw`]
+   * [`u32`:`commitment_feerate_perkw`]
+   * [`u64`:`funding_satoshis`]
+   * [`u64`:`dust_limit_satoshis`]
+   * [`u64`:`max_htlc_value_in_flight_msat`]
+   * [`u64`:`htlc_minimum_msat`]
+   * [`u16`:`to_self_delay`]
+   * [`u16`:`max_accepted_htlcs`]
+   * [`u32`:`locktime`]
+   * [`point`:`funding_pubkey`]
+   * [`point`:`revocation_basepoint`]
+   * [`point`:`payment_basepoint`]
+   * [`point`:`delayed_payment_basepoint`]
+   * [`point`:`htlc_basepoint`]
+   * [`point`:`first_per_commitment_point`]
+   * [`byte`:`channel_flags`]
+   * [`opening_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `opening_tlvs`
```

Comment:
Can we also add an empty tlv to both `open_channel2` and `accept_channel2` that signals that a node will only accept confirmed inputs? Something like (I'm not sure how to define an empty tlv without breaking your spec parsing):

```
1. type: 2 (`confirmed_inputs_only`)
2. data:
    * []
```

Only accepting confirmed inputs ensures that the funding tx doesn't end up over-paying fees to make up for low-feerate unconfirmed ancestors of your counterparty. This can be quite useful for LSPs who end up opening a lot of dual-funded channels with wallet users and want to make sure they don't get "robbed" on on-chain fees.